### ISSUE TITLE
Tensor To OpenCV - C++

### ISSUE THREAD

**Issue by pedromorais17** [pytorch/pytorch#20011]
I'm using the pytorch with float image (C++). Currently, I'm not using GPU, but I'm having difficulties to convert the data between torch and opencv. 

Basically, I'm reading an image,

cv::Mat image, image_bgr;
image_bgr = cv::imread("FILENAME");
cvtColor(image_bgr, image, cv::COLOR_BGR2RGB);

// converting it to float
cv::Mat bgr[3]; // destination array
cv::split(image, bgr);
cv::Mat channelsConcatenated;
vconcat(bgr[0], bgr[1], channelsConcatenated);
vconcat(channelsConcatenated, bgr[2], channelsConcatenated);
cv::Mat channelsConcatenatedFloat;
channelsConcatenated.convertTo(channelsConcatenatedFloat, 21);

std::vector<int64_t> dims{ 1, static_cast<int64_t>(image.channels()),
							  static_cast<int64_t>(image.rows),
							  static_cast<int64_t>(image.cols) };

// creating my tensor
auto options = torch::TensorOptions().dtype(torch::kFloat).layout(torch::kStrided).device(torch::kCPU);
at::Tensor tensor_image =
		torch::from_blob(channelsConcatenatedFloat.data, at::IntList(dims),options);

// Then I'm performing the inverse transform, and I want to see the image in an opencv window. 
 cv::Mat testTensorInit(cv::Size(700, 395), 21,tensor_image.clone().squeeze(0).to(at::kByte).data<float>());

But it always crashing. Do you have any suggestion?

If convert to uint8_t it works, however the image result is not correct due to the number of bytes. 
Please see the result. I'm trying to solve the issue, but the available information is limited. 

![veins](https://user-images.githubusercontent.com/44382268/57025917-6594ee80-6c30-11e9-9095-4acb9d16cafc.PNG)

**Comment by xsacha** [comment]
Something like this?
```C++
    // Convert a vector of images to torch Tensor
    torch::Tensor matToTensor(const cv::Mat &image)
    {
        bool isChar = (image.type() & 0xF) < 2;
        std::vector<int64_t> dims = {image.rows, image.cols, image.channels()};
        return torch::from_blob(image.data, dims, isChar ? torch::kChar : torch::kFloat).to(torch::kFloat);
    }

    cv::Mat matToTensor(const torch::Tensor &tensor)
    {
        auto sizes = tensor.sizes();
        return cv::Mat{cv::Size{sizes[0], sizes[1]}, CV_32FC(sizes[2]), tensor.data<float>());
    }
```

**Comment by zhangguanheng66** [comment]
Please use https://discuss.pytorch.org for questions. Please kindly feel encouraged to reopen this issue, if this does not apply.

**Comment by pedromorais17** [comment]
Dear all,

I'm not sure if you have a GPU implementation example. Looking for the previous example, the CPU code is working correctly, but the GPU example is failing completely. Please see the example code. Have any suggestion? Thank you in advance. 

// read image
cv::Mat image_bgr;
cv::Mat image;
image_bgr = cv::imread("C:\\Users\\pedro\\Desktop\\binary\\VeinsImages\\teste_0.png");
cv::cvtColor(image_bgr, image, cv::COLOR_BGR2RGB);

//normalize
cv::Mat imageFloat;
image.convertTo(imageFloat, 21);
normalize(imageFloat, imageFloat, 0, 1, 32);

//send to gpu
cv::cuda::GpuMat gpuImage;
gpuImage.upload(imageFloat);

std::vector<int64_t> sizes = { 1,  static_cast<int64_t>(gpuImage.rows),				 static_cast<int64_t>(gpuImage.cols), static_cast<int64_t>(gpuImage.channels())};

// create tensor
auto options = torch::TensorOptions().dtype(torch::kFloat32).device(torch::kCUDA);
at::Tensor tensor_image =torch::from_blob(gpuImage.data, torch::IntList(sizes),options);

// permute dimensions
tensor_image = tensor_image.permute({ 0, 3, 1, 2 });

//load module
std::shared_ptr<torch::jit::script::Module> module = torch::jit::load(FILENAME);

//get output and convert it to opencv binary image
auto output = module->forward(input).toTensor();
cv::Mat outputData(cv::Size(700, 395), 0, output.toBackend(torch::Backend::CPU).clone().squeeze(0).to(at::kByte).data<uint8_t>());

**Comment by xsacha** [comment]
For GPU, you must use from_blob_strided as the GPU mat will be strided.
Strides will be {step, channels, 1}

**Comment by pedromorais17** [comment]
Thank you @xsacha. Indeed, I saw an example using strides, but I failed to set correctly the dimensions. 
However, how can I reconvert from a strided tensor to a cvMat structure?

Thank you in advance.

**Comment by pedromorais17** [comment]
I found the solution. I'm sending an example of the conversion step to a binary image:

```
 output = output.permute({ 0, 2, 3, 1 });
 output = output.squeeze(0).detach();
 output = output.mul(255).clamp(0, 255).to(torch::kU8);
 output = output.to(torch::kCPU);
 cv::Mat resultImg(395, 700, 0);
 std::memcpy((void*)resultImg.data, output.data_ptr(), sizeof(torch::kU8)*output.numel());
```

**Comment by xsacha** [comment]
Woah, what are you doing there? That's a lot of modifying the image.
I understand the permute if it's the result of a forward, but why the multiple/clamp? If the result is floating point it is probably for a reason. Keep it in 32F?

You can also dump in straight in to a GpuMat. Then run a convertTo.

**Comment by pedromorais17** [comment]
Yes @xsacha , I'm receving FLOAT data and converting to unsigned char. Indeed, your suggestion is more efficient, I will update it. Thank you very much for your support.

**Comment by pedromorais17** [comment]
Dear @xsacha, how can I directly convert a tensor to a gpumat. I'm using the following code:


```
auto output = m_cudaModule->forward(input).toTensor();
output = output.permute({ 0, 2, 3, 1 });
output = output.squeeze(0).detach();
output = output.mul(255).to(torch::kU8);

cv::cuda::GpuMat gMat = cv::cuda::GpuMat(cv::Size(m_processedImage.size()), CV_8UC1, output.data<unsigned char>());

```

I'm having problem when accessing my gMat variable.  Do you have any suggestion?

**Comment by xsacha** [comment]
To be honest, I haven't needed that yet. My outputs are always vectors that I want on CPU.

**Comment by xsidneib** [comment]
This worked for me:
From cv::cuda::GpuMat to tensor (in cuda). 
`
torch::Tensor to_tensor(const cv::cuda::GpuMat& im) {
  std::vector<int64_t> dims = {1, im.channels(), im.rows, im.cols};
  int64 step = im.step;// / sizeof(float);
  std::vector<int64_t> strides = {1, 1, step, im.channels()};
  auto options = torch::TensorOptions().dtype(torch::kUInt8).device(torch::kCUDA);
  return torch::from_blob(im.data, dims, strides,  options);
}
`
And converting to cv::Mat.
`
cv::Mat to_mat(const at::Tensor& tensor) {
  auto output = tensor.permute({ 0, 2, 3, 1 });
  output = output.squeeze(0).detach();
  output = output.to(torch::kU8).to(torch::kCPU);

  auto sizes = output.sizes();
  cv::Mat mat{cv::Size{static_cast<int>(sizes[1]) ,
                       static_cast<int>(sizes[0]) },
              CV_8UC(static_cast<int>(sizes[2])),
              output.data_ptr()};

//  cv::imwrite("/tmp/mat.png", mat);
  return mat;
}
`

**Comment by DineshRajanT** [comment]
> I'm using the pytorch with float image (C++). Currently, I'm not using GPU, but I'm having difficulties to convert the data between torch and opencv.
> 
> Basically, I'm reading an image,
> 
> cv::Mat image, image_bgr;
> image_bgr = cv::imread("FILENAME");
> cvtColor(image_bgr, image, cv::COLOR_BGR2RGB);
> 
> // converting it to float
> cv::Mat bgr[3]; // destination array
> cv::split(image, bgr);
> cv::Mat channelsConcatenated;
> vconcat(bgr[0], bgr[1], channelsConcatenated);
> vconcat(channelsConcatenated, bgr[2], channelsConcatenated);
> cv::Mat channelsConcatenatedFloat;
> channelsConcatenated.convertTo(channelsConcatenatedFloat, 21);
> 
> std::vector<int64_t> dims{ 1, static_cast<int64_t>(image.channels()),
> static_cast<int64_t>(image.rows),
> static_cast<int64_t>(image.cols) };
> 
> // creating my tensor
> auto options = torch::TensorOptions().dtype(torch::kFloat).layout(torch::kStrided).device(torch::kCPU);
> at::Tensor tensor_image =
> torch::from_blob(channelsConcatenatedFloat.data, at::IntList(dims),options);
> 
> // Then I'm performing the inverse transform, and I want to see the image in an opencv window.
> cv::Mat testTensorInit(cv::Size(700, 395), 21,tensor_image.clone().squeeze(0).to(at::kByte).data());
> 
> But it always crashing. Do you have any suggestion?
> 
> If convert to uint8_t it works, however the image result is not correct due to the number of bytes.
> Please see the result. I'm trying to solve the issue, but the available information is limited.
> 
> ![veins](https://user-images.githubusercontent.com/44382268/57025917-6594ee80-6c30-11e9-9095-4acb9d16cafc.PNG)

`output = output.permute({ 0, 2, 3, 1 });
	output = output.squeeze(0).detach();
	output = output.mul(255).clamp(0, 255).to(torch::kU8);
	auto sizes = output.sizes();
	cout << "sizes : " << sizes << endl;
	output = output.to(torch::kCPU);
	cout << "CV_8UC()---> :"<<static_cast<int>(sizes[2]) << endl;
	//cv::Mat resultImg(static_cast<int>(sizes[1]),static_cast<int>(sizes[0]), 3);
	//std::memcpy((void*)resultImg.data, output.data_ptr(), sizeof(torch::kU8)*output.numel());
	cv::Mat mat{cv::Size{static_cast<int>(sizes[1]) ,	
						 static_cast<int>(sizes[0]) },
				CV_8UC(static_cast<int>(sizes[2])),
				output.data_ptr() };
	//cv::cvtColor(mat, mat, cv::COLOR_GRAY2RGB);
	cout << "Saving image";
	cv::imwrite("Y:/back_ups/phospheneai/Images/resultImg_final.jpg", output);`

I am trying this code to save the inferred image.But I get the same error as yours @pedromorais17 .I get grayscale grid of 9 images. Did something work for you?

**Comment by xsacha** [comment]
This has been solved already in the comments.

**Comment by Yuwei1Tian** [comment]
> Dear @xsacha, how can I directly convert a tensor to a gpumat. I'm using the following code:
> 
> ```
> auto output = m_cudaModule->forward(input).toTensor();
> output = output.permute({ 0, 2, 3, 1 });
> output = output.squeeze(0).detach();
> output = output.mul(255).to(torch::kU8);
> 
> cv::cuda::GpuMat gMat = cv::cuda::GpuMat(cv::Size(m_processedImage.size()), CV_8UC1, output.data<unsigned char>());
> ```
> 
> I'm having problem when accessing my gMat variable. Do you have any suggestion?

Have you solved the problem yet? I tried to use cudaMemcpy to convert tensor directly to GpuMat, but only it only works for some conditions.  For inputs of 2560*2560, it will return a fine result, however, other resolutions don't work.
"cudaMemcpy(gMat.data, output.data_ptr(), sizeof(torch::kU8)*t_out.numel(), cudaMemcpyDevicetoDevice); "

**Comment by xsacha** [comment]
The solution provided above does exact what you're asking for.

```
torch::Tensor to_tensor(const cv::cuda::GpuMat& im) {
  std::vector<int64_t> dims = {1, im.channels(), im.rows, im.cols};
  int64 step = im.step;// / sizeof(float);
  std::vector<int64_t> strides = {1, 1, step, im.channels()};
  auto options = torch::TensorOptions().dtype(torch::kUInt8).device(torch::kCUDA);
  return torch::from_blob(im.data, dims, strides,  options);
}
```

**Comment by Yuwei1Tian** [comment]
> You need to account for the pitch/stride of the cuda tensor. A size of 2560 may also have a stride of 2560, which is why it worked for you. Create a GpuMat with the same stride of the torch tensor instead. Then, if you copy (don't need to), use the stride instead of width.

Thank you, I'll have a try
